draft-chen-cfrg-vdaf-pine.md

```diff
@@ -116,122 +123,104 @@ the aggregation step across multiple servers such that no server sees any
-{{Section 6.1 of !VDAF}}. Let `q` denote the field size. In our case, the
-circuit would compute the sum of square of all entries, i.e. squared L2-norm,
-modulo field size `q`, and check the result is in the desired range.
-Note we don't take the square root, because it's more efficient and
-mathematically equivalent to verify the result in squared L2-norm.
-
-However, it is challenging to bound each field element in the vector, which
-allows a malicious Client to send a vector that causes the computation of sum
-of squares to overflow, or "wrap around" field size. When Aggregators evaluate
-this computation over secret shares of field integer vector, there is no way
-for the Aggregators to determine if wraparound has happened, and Aggregators
-may accept the vector from the malicious Client, which causes the final
-aggregate to be corrupted.
-
-One way to resolve this issue is to encode each coordinate of the vector
-as its binary representation, send each bit as an individual field element,
-and enforce a dimension limit on the Client vector.
-By checking each field element is indeed a bit, the Aggregators can naturally
-bound each vector entry, and with a dimension limit, Aggregators can ensure
-that computing sum of squares won't wrap around field size `q`. However,
-this approach results in a significant amount of communication overhead
-between Client and Aggregators, because the Client needs to send at least
-`num_frac_bits * dimension` field elements, with `num_frac_bits` and
-`dimension` specified by users in {{pine-user-param}}.
+One way to achieve this is to use an FLP ("Fully Linear Proof"; see {{Section
+7.3 of !VDAF}}). An FLP is an assertion about the validity of some input that
+can be checked jointly by the Aggregators who only hold secret shares of the
+input. Validity is expressed as a circuit evaluated on the input over a finite
+field, denoted by `Field` in {{Section 6.1 of !VDAF}}. Let `q` denote the field
+size. In our case, the circuit would compute the sum of the squares of the
+entries of the gradient vector, that is the squared L2-norm, and check that the
+result is in the desired range. Note that there is no need to compute the exact
+L2-norm (i.e., the square root of the sum); it suffices to compute the squared
+L2-norm and check that it is smaller than the square of the bound.
+
+Crucially, arithemtic in this computation is modulo the field size `q`. This
+means that, for a given gradient, the norm may have a different result when
+computed in our finite field versus the ring of integers. For example, suppose
+our bound is `10`: the gradient `[99, 0, 7]` has squared L2-norm of `9850` over
+the integers (out of range), but only `6` modulo `q = 23` (in range). This is
+becuase the sum of the squares "wraps around" the field modulus `q`.
+
+Thus the central challenge of adapting FLPs to this problem is preventing such
+"wraparounds".
+
+One way to do this is to encode each entry of the gradient such that the
```

Comment:
I think it's quick enough to just mention "one way to do this is to encode the bit representation of each entry of the gradient such that..."? Otherwise it's not immediately clear why the communication cost is `O(num_frac_bits * dimension)`?